File: src/com/sterndu/data/transfer/Packet.java
package com.sterndu.data.transfer;

import java.util.Arrays;
import java.util.Objects;

public final class Packet {
	private final byte type;
	private final byte[] data;

	public Packet(byte type, byte[] data) {
		this.type = type;
		this.data = data;
	}

	public byte type() {
		return type;
	}

	public byte[] data() {
		return data;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == this) return true;
		if (obj == null || obj.getClass() != this.getClass()) return false;
		Packet that = (Packet) obj;
		return this.type == that.type &&
				Arrays.equals(this.data, that.data);
	}

	@Override
	public int hashCode() {
		return Objects.hash(type, Arrays.hashCode(data));
	}

	@Override
	public String toString() {
		return "Packet[" +
				"type=" + type + ", " +
				"data=" + Arrays.toString(data) + ']';
	}

}

File: src/com/sterndu/data/transfer/Packet.kt
@file:JvmName("Packet")
package com.sterndu.data.transfer

import java.util.*

data class Packet(val type: Byte, val data: ByteArray) {

	override fun equals(other: Any?): Boolean {
		if (other === this) return true
		if (other == null || other.javaClass != this.javaClass) return false
		val that = other as Packet
		return type == that.type &&
				data.contentEquals(that.data)
	}

	override fun hashCode(): Int {
		return Objects.hash(type, data.contentHashCode())
	}

	override fun toString(): String {
		return "Packet[" +
				"type=" + type + ", " +
				"data=" + data.contentToString() + ']'
	}
}
